vnext/Microsoft.ReactNative/CoalescingEventEmitter.h

```diff
@@ -0,0 +1,49 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include "CoalescingEventEmitter.g.h"
+#include "JSValue.h"
+#include "ReactHost/React.h"
+#include "ReactPropertyBag.h"
+#include "winrt/Microsoft.ReactNative.h"
+
+#include <deque>
+#include <mutex>
+
+namespace winrt::Microsoft::ReactNative::implementation {
+
+struct CoalescingEvent
+{
+  std::optional<int64_t> coalesceKey;
+  bool expired{false};
+  folly::dynamic args;
+};
+
+struct CoalescingEventEmitter : CoalescingEventEmitterT<CoalescingEventEmitter> {
+ public:
+  static ReactPropertyId<CoalescingEventEmitter> Property() noexcept;
+  static winrt::Microsoft::ReactNative::CoalescingEventEmitter FromContext(IReactContext const & context) noexcept;
+  static winrt::Microsoft::ReactNative::CoalescingEventEmitter FromContext(Mso::React::IReactContext const & context) noexcept;
+
+  CoalescingEventEmitter(Mso::CntPtr<Mso::React::IReactContext> &&context) noexcept;
+
+  void EmitJSEvent(int64_t coalesceKey, JSValueArgWriter const & paramsArgWriter) noexcept;
+  void EmitDurableJSEvent(JSValueArgWriter const & paramsArgWriter) noexcept;
+
+ private:
+  void EnsureRenderCallbackRegistered() noexcept;
+  void OnRendering() noexcept;
+
+  Mso::CntPtr<Mso::React::IReactContext> m_context;
+  std::deque<CoalescingEvent> m_eventQueue;
```

Comment:
This could grow unboundedly? consider using a fixed max size circular buffer, since we probably don't care about events that are "too old" anyway?

Reply:
plus dequeue is non contiguous and allocs, so it's not great for perf

Reply:
dequeue is a single contiguous circular buffer (that re-allocs when full). We definitely do care about reliable delivery of events, so I do not think bounding makes sense.

Reply:
https://en.cppreference.com/w/cpp/container/deque
_As opposed to std::vector, the elements of a deque are not stored contiguously: typical implementations use a sequence of individually allocated fixed-size arrays, with additional bookkeeping, which means indexed access to deque must perform two pointer dereferences, compared to vector's indexed access which performs only one._

https://github.com/microsoft/STL/blob/b52c3797e262cbce28166d92f9d1b8f2bfeb5ac5/stl/inc/deque#L568

Reply:
Interesting. I knew deque was buffer based but didn't realize it used multiple buffers.